### DISCUSSION TITLE
MPC Implementing Real System Measurements

### DISCUSSION THREAD

**Discussion by sherlock-ed-hub** [do-mpc/do-mpc#209] [Q&A] [answered]
Hi,
first things first: Great Work!
I implemented the ODE from here: https://apmonitor.com/do/index.php/Main/DynamicControl and adapted the variables no problem everything works fine.

But now I try to implement my real system where I try to input my actual measurements. Where and how do I do it?
I tried commenting out the simulator in the control loop and put the measurements into the x0 array. But I guess that's not all I have to do. And I don't know how to calculate the differential v'. Because it doesn't seem to be the slope of the last two points in time. 

Maybe I just missed where to put the measurements and how to calculate all that is not directly measurable. If you could help me that would be great :)

**Comment by 4flixt** [comment, marked as answer]
Hi @sherlock-ed-hub ,

thank you for the compliment and for using do-mpc! Also apologies for the late reply ...

Regarding your question: The main-loop in do-mpc looks something like this:

```python
# Initial condition
x = ....

while [condition]:
    u = mpc.make_step(x)
    y = simulator.make_step(u)
    x = estimator.make_step(y)
```

Note that ``x,u,y`` are all ``numpy.ndarray``  of appropriate dimension. If you want to now substitute the Simulator for the true plant the only thing you have to do is to pass the input ``u`` (somehow) to your plant and retrieve the measurements ``y``. It is that simple. Of course the communication with your plant may be arbitrarily complex but after you have managed that, the integration in do-mpc shouldn't be a problem.

Best
Felix

**Reply by edumagol** [reply]
I had the same doubt @4flixt ! Thanks for your repply. I will try to do it and give a feedback to you ASAP.

By the way, amazing repo do-mpc... probably one of the few in the world about MPC using python!
